readme.txt
lavagimp v2.0.0-a3 Development version (C) lavajoe 2018 see the LICENSE file for details regarding the GNU GPL v.3 license.
The latest soure code for lavagimp, as well as the libconfig 1.5 library it uses is availble here: https://github.com/lavajoe/lavagimp

This is the most current sourecode of lavagimp (with the source for libconfig 1.5 that I installed/build lavagimp with) with an included
complete binary build thats the latest one I've compiled, please be careful and expect bugs when using the latest build as indicated by the
version it could be big update.

See changelog

To use a different version of libconfig or your own modified version with lavagimp, you can replace the libconfig-9.dll in lavagimp.exe.local
with the modified/newer version, or rebuild lavagimp from source (availble from the github link above) to use the newer/modified version.

Usage:

By default lavagimp will load gimps.txt from the gimps folder in the .exes directory, the default hotkeys are insert to autotype a gimp,
home to switch modes from regular autotype mode, slow type mode which will add a delay between keystrokes, top to bottom mode and
gimp collection mode which takes whatevers in your clipboard and puts it at the end of the loaded gimpfile, to configure lavagimp to
change the hotkeys/delay in keystrokes in slow type mode see the config.cfg in the config folder of the .exes directory.

changelog.txt
v2.0.0-a3 (pre-release current development):
	Added the ability to add a phrase before or after a gimp in config
	Added the ability to auto press enter or not in config
	Added the ability to set time between keypresses in slow type mode in config
	Check for errors registering the hotkeys when lavagimp starts, will warn you if another instance of lavagimp is already running
v2.0.0-a2 (pre-release current development):
	Added a version number for development builds, the first build for lavagimp 2.0.0 is a1, the current one as of now is a2, sorry for
	any confusion.
	Added the ability to change all the current hotkeys (insert gimp, change mode, delete), right now its just a few keys but more are coming soon.
v2.0.0-a1 (pre-release current development):
	Added gimp directory (default location of gimps.txt is gimps\gimps.txt, can change it to anything you want)
	Added config directory which contains a config.cfg file to config lavagimp, current options are change home key to a few different keys
	and change gimp file location.
	Added ability to change hotkeys, only like 6 currently, not sure if I should add every single key on keyboard or just the ones that make sense
	currently have most of the ones that make sense (to me atleast)
v1.3.3:
	Fix random characters being added at the end of gimps
v1.3.2:
	inital release

keys.c
/*
    lavagimp, autotype random quotes (gimps) by pressing hotkeys
	
    Copyright (C) 2018  lavajoe <joe89gunnar@gmail.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include "lavagimp.h"

UINT findKey(char key[64]) {
	
	UINT modekey = 69696969;

	if (strcmp("home", key) == 0) {
	  modekey = 0x24;
	}
	if (strcmp("delete", key) == 0) {
	  modekey = 0x2E;
	}
	if (strcmp("insert", key) == 0) {
	  modekey = 0x2D;
	}
	if (strcmp("page_up", key) == 0) {
	  modekey = 0x21;
	}
	if (strcmp("page_down", key) == 0) {
	  modekey = 0x22;
	}
	if (strcmp("end", key) == 0) {
	  modekey = 0x23;
	}
	if (modekey == 69696969) {
		return 69696969;
	}

	return modekey;
}



main.c
/*
    lavagimp, autotype random quotes (gimps) by pressing hotkeys
	
    Copyright (C) 2018  lavajoe <joe89gunnar@gmail.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <windows.h>
#include <ctype.h>
#include <unistd.h>
#include <conio.h>
#include <stdbool.h>

#include <libconfig.h>

char gimp[999999][512];
int gimpN = 0;

int* allocate_mem(int*** arr, int n, int m)
{
  *arr = (int**)malloc(n * sizeof(int*));
  int *arr_data = malloc( n * m * sizeof(int));
  for(int i=0; i<n; i++)
     (*arr)[i] = arr_data + i * m ;
  return arr_data; //free point
}

void deallocate_mem(int*** arr, int* arr_data){
    free(arr_data);
    free(*arr);
}

int fileExists(char *filename) {
	if(access(filename, F_OK) != -1) {
    	return 1;
	} else {
	    return 0;
	}
}

void BailOut(char *msg)
{
	fprintf(stderr, "Exiting: %s\n", msg);
	exit(1);
}

int main() {

	printf("lavagimp version 2.0.0. Copyright (C) 2018 lavajoe\nThis program comes with ABSOLUTELY NO WARRANTY; for details see the 'LICENSE' file.\nThis is free software, and you are welcome to redistribute it under certain conditions;\nsee the 'LICENSE' file for details.\n\n");
	
	//Load config file
	config_t cfg;
	
	const char *gimpfilecfg;
	config_init(&cfg);
	const char *beforegimp;
	const char *aftergimp;
	UINT modekey;
	UINT insertkey;
	UINT deletegimp;
	bool pbg = false;
	bool pag = false;
	int putenter;
	int sleeptime;
	if (config_read_file(&cfg, "config/config.cfg") == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	

	if (config_lookup_string(&cfg, "gimpfile", &gimpfilecfg) == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	if (config_lookup_int(&cfg, "insertgimp", &insertkey) == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	if (config_lookup_int(&cfg, "changemode", &modekey) == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	if (config_lookup_int(&cfg, "deletegimp", &deletegimp) == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	if (config_lookup_string(&cfg, "beforegimp", &beforegimp) == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	if (config_lookup_string(&cfg, "aftergimp", &aftergimp) == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	if (config_lookup_bool(&cfg, "putenter", &putenter) == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	if (config_lookup_int(&cfg, "sleeptime", &sleeptime) == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}

	if (strlen(beforegimp) < 1) {
		pbg = false;
	} else {
		pbg = true;
	}
	if (strlen(aftergimp) < 1) {
		pag = false;
	} else {
		pag = true;
	}
		
	//printf("beforegimp: %s.\naftergimp: %s.\n",beforegimp,aftergimp);
	//UINT modekey = modekey2;
	char gimpfile[64];
	strcpy(gimpfile, gimpfilecfg);


	printf("Loaded gimplist from: %s.\n",gimpfile);
	
	FILE *GIMPS;
	if (fileExists(gimpfile)) {
		char gimpBuffer[512];
		gimpN = 1;
		GIMPS = fopen(gimpfile, "r");
		char ch;
		while(!feof(GIMPS)) {
  			ch = fgetc(GIMPS);
  			if(ch == '\n') {
    			gimpN++;
  			}
  		}

  		//fclose(GIMPS);
	  	GIMPS = fopen(gimpfile, "r");
	  	int i = 1;
	  	for (i = 1; i <= gimpN; i++) {
			fgets(gimp[i], sizeof(gimp[i]), GIMPS);
			memset(gimp[i]+strlen(gimp[i])-1, '\0', 1);
		}
	} else {
		printf("gimps.txt not found.\n");
		getchar();
		exit(1);
		gimpN = -1;
	}

	if (gimpN < 1) {
		printf("gimps.txt is empty!");
		getchar();
		exit(1);
	}

	//RegisterHotKey(NULL, 100, MOD_ALT | MOD_CONTROL, 'S');

	UINT hk = NULL;
	HKL kl = LoadKeyboardLayout("US", KLF_ACTIVATE);
	if (RegisterHotKey(NULL, 1, hk, insertkey) == 0) {
		printf("Failed to register insert hotkey! (Maybe you have another instance of lavagimp open)\n");
		getchar();
		exit(1);
	}
	if (RegisterHotKey(NULL, 2, hk, modekey) == 0) {
		printf("Failed to register modechange hotkey! (Maybe you have another instance of lavagimp open)\n");
		getchar();
		exit(1);
	}
	if (RegisterHotKey(NULL, 3, hk, deletegimp) == 0) {
		printf("Failed to register insert hotkey! (Maybe you have another instance of lavagimp open)\n");
		getchar();
		exit(1);
	}
	RegisterHotKey(NULL, 4, hk, 0xA5);
	

	MSG msg;
	srand(time(NULL));

	int delLength = 0;
	int gimpLog = 1;
	int mode = 1;
	/*char htest;
        printf("enter: ");
	scanf("%c", &htest);
	//short vret = VkKeyScanEx('I', kl);
	printf("\ngot '%hd'",vret);*/
	/*
	HANDLE h;
   
  if (!OpenClipboard(NULL))
    BailOut("Can't open clipboard");
   
  h = GetClipboardData(CF_TEXT);
 
  printf("%s\n", (char *)h);
 
  CloseClipboard();*/

	while (GetMessageA(&msg, NULL, 0, 0) != 0) {//main loop
		if (msg.wParam == 3) {//delete gimp (end key)
			int i = 0;
			for (i = 0; i < delLength + 1; i++) {
				keybd_event(0x08,0,0,0);
				keybd_event(0x08,0,KEYEVENTF_KEYUP,0);
			}
		}
		if (msg.wParam == 2) {//mode switch (home key)
			mode++;
			if (mode == 5) {
				mode = 1;
			}
			switch (mode) {
				case 1:
					printf("Switched to: normal random mode (1)\n");
					break;
				case 2:
					printf("Switched to: slow type mode (2)\n");
					break;
				case 3:
					printf("Switched to: top to bottom mode (3)\n");
					break;
				case 4:
					printf("Switched to: Gimp collection mode (4)\n");
					break;
			}
		}
		if (msg.wParam == 1) {//action (insert key)
			if (mode == 4) {//gimp collection mode
				
				HANDLE h;
   
				if (!OpenClipboard(NULL)) {
					BailOut("Can't open clipboard");
				}
				h = GetClipboardData(CF_TEXT);
				printf("Putting gimp \"%s\" into gimps.txt.\n", (char *)h);
				CloseClipboard();
				
				/*GIMPS = fopen("gimps.txt", "w");
				fprintf(GIMPS, "%s",(char*)h);
				fclose(GIMPS);
				FILE *GIMPS2;
				int gimpN2;
				*/
				FILE *GIMPS2;
				//int gimpN2;
				GIMPS2 = fopen(gimpfile, "a+");
				//char gimp2[999999][512];
				
				if (!fileExists(gimpfile)) {	
					printf("gimps.txt not found!\n");
					exit(1);
				}
				
				fprintf(GIMPS2, "\n%s", (char*)h);
				fclose(GIMPS2);
				
			}
			if (mode == 3) {//start to bottom mode
				 int i = 0;
				 char buf[512];
				 memset(buf, 0, 512);
				 if (pbg) {
					strcat(buf, beforegimp);
				 }
				 strcat(buf, gimp[gimpLog]);
				 if (pag) {
					strcat(buf, aftergimp);
				 }
				 delLength = strlen(buf);
				 for (i = 0; i < strlen(buf); i++) {
					if (buf[i] == '?' || buf[i] == '>' || buf[i] == ':' || buf[i] == '!' || buf[i] == '@' || buf[i] == '#' || buf[i] == '$' || buf[i] == '%' || buf[i] == '^' || buf[i] == '&' || buf[i] == '*' || buf[i] == '(' || buf[i] == ')' || buf[i] == '_' || buf[i] == '+' || buf[i] == '"' || buf[i] == '<' || buf[i] == '|' || buf[i] == '~' || isupper(buf[i]) != 0) {
						keybd_event(0x10,0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
						keybd_event(0x10,0,KEYEVENTF_KEYUP,0);
					} else {
						keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
					}
				 }
				 if (putenter) {
					 keybd_event(0x0D,0,0,0);
				 }
				 
				if (gimpLog == gimpN) { gimpLog = 0; }
				gimpLog++;
			}
			if (mode == 2) { //slow type
				 int randomnumber = rand() % (gimpN - 1) + 1;
				 int i = 0;
				 char buf[512];
				 memset(buf, 0, 512);
				 if (pbg) {
					strcat(buf, beforegimp);
				 }
				 strcat(buf, gimp[randomnumber]);
				 if (pag) {
					strcat(buf, aftergimp);
				 }
				 delLength = strlen(buf);
				 for (i = 0; i < strlen(buf); i++) {
					if (buf[i] == '?' || buf[i] == '>' || buf[i] == ':' || buf[i] == '!' || buf[i] == '@' || buf[i] == '#' || buf[i] == '$' || buf[i] == '%' || buf[i] == '^' || buf[i] == '&' || buf[i] == '*' || buf[i] == '(' || buf[i] == ')' || buf[i] == '_' || buf[i] == '+' || buf[i] == '"' || buf[i] == '<' || buf[i] == '|' || buf[i] == '~' || isupper(buf[i]) != 0) {
						keybd_event(0x10,0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
						keybd_event(0x10,0,KEYEVENTF_KEYUP,0);
					} else {
						keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
					}
					Sleep(sleeptime); //60 before, 90 after, 75 now?
				 }
				 if (putenter) {
					 keybd_event(0x0D,0,0,0);
				 }
			}
			if (mode == 1) {//regular
				int randomnumber = rand() % (gimpN - 1) + 1;
             			int i = 0;
             			char buf[512];
				memset(buf, 0, 512);
				if (pbg) {
					strcat(buf, beforegimp);
				}
				//strncpy(buf, gimp[randomnumber], strlen(gimp[randomnumber]));
				strcat(buf, gimp[randomnumber]);
				if (pag) {
					strcat(buf, aftergimp);
				}
             			delLength = strlen(buf);
             			for (i = 0; i < strlen(buf); i++) {
             				if (buf[i] == '?' || buf[i] == '>' || buf[i] == ':' || buf[i] == '!' || buf[i] == '@' || buf[i] == '#' || buf[i] == '$' || buf[i] == '%' || buf[i] == '^' || buf[i] == '&' || buf[i] == '*' || buf[i] == '(' || buf[i] == ')' || buf[i] == '_' || buf[i] == '+' || buf[i] == '"' || buf[i] == '<' || buf[i] == '|' || buf[i] == '~' || isupper(buf[i]) != 0) {
             					keybd_event(0x10,0,0,0);
             					keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
             					keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
             					keybd_event(0x10,0,KEYEVENTF_KEYUP,0);
             				} else {
             					keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
             					keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
             				}
              			}
				if (putenter) {
					 keybd_event(0x0D,0,0,0);
				}
	         		//keybd_event(VkKeyScanEx(' ', kl),0,0,0); //print space at the end of gimp
             			//keybd_event(VkKeyScanEx(' ', kl),0,KEYEVENTF_KEYUP,0);

			}
		}

		//delLength = strlen(gimp[randomnumber]);
	}

	return 0;
} 


lavagimp.h
/*
    lavagimp, autotype random quotes (gimps) by pressing hotkeys
	
    Copyright (C) 2018  lavajoe <joe89gunnar@gmail.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <windows.h>

UINT findKey(char key[64]);



lavagimp.c
/*
	lavagimp, autotype random quotes (gimps) by pressing hotkeys
	
    Copyright (C) 2018  lavajoe <joe89gunnar@gmail.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <windows.h>
#include <ctype.h>
#include <unistd.h>
#include <conio.h>

#include <libconfig.h>

char gimp[999999][512];
int gimpN = 0;

int* allocate_mem(int*** arr, int n, int m)
{
  *arr = (int**)malloc(n * sizeof(int*));
  int *arr_data = malloc( n * m * sizeof(int));
  for(int i=0; i<n; i++)
     (*arr)[i] = arr_data + i * m ;
  return arr_data; //free point
}

void deallocate_mem(int*** arr, int* arr_data){
    free(arr_data);
    free(*arr);
}

int fileExists(char *filename) {
	if(access(filename, F_OK) != -1) {
    	return 1;
	} else {
	    return 0;
	}
}

void BailOut(char *msg)
{
	fprintf(stderr, "Exiting: %s\n", msg);
	exit(1);
}

int main() {

	printf("lavagimp version 2.0.0 (development version). Copyright (C) 2018 lavajoe\nThis program comes with ABSOLUTELY NO WARRANTY; for details see the 'LICENSE' file.\nThis is free software, and you are welcome to redistribute it under certain conditions;\nsee the 'LICENSE' file for details.\n\n");
	
	//Load config file

	config_t cfg;
	
	const char *gimpfilecfg;
	config_init(&cfg);
	const char *modekey2;
	if (config_read_file(&cfg, "config/config.cfg") == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	

	if (config_lookup_string(&cfg, "gimpfile", &gimpfilecfg) == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	if (config_lookup_string(&cfg, "changemode", &modekey2) == CONFIG_FALSE) {
		printf("Error parsing config file: %s.\n",config_error_text(&cfg));
	}
	//UINT modekey = modekey2;
	char gimpfile[64];
	strcpy(gimpfile, gimpfilecfg);
	char modekey3[64];
	strcpy(modekey3, modekey2);
	UINT modekey = 69696969;

	if (strcmp("home", modekey3) == 0) {
	  modekey = 0x24;
	}
	if (strcmp("delete", modekey3) == 0) {
	  modekey = 0x2E;
	}
	if (strcmp("insert", modekey3) == 0) {
	  modekey = 0x2D;
	}
	if (strcmp("page_up", modekey3) == 0) {
	  modekey = 0x21;
	}
	if (strcmp("page_down", modekey3) == 0) {
	  modekey = 0x22;
	}
	if (strcmp("end", modekey3) == 0) {
	  modekey = 0x23;
	}
	if (modekey == 69696969) {
	  printf("Invalid hotkey: %s\n",modekey3);
	  getchar();
	  exit(1);
	}

	printf("Loaded gimplist from: %s.\n",gimpfile);
	
	FILE *GIMPS;
	if (fileExists(gimpfile)) {
		char gimpBuffer[512];
		gimpN = 1;
		GIMPS = fopen(gimpfile, "r");
		char ch;
		while(!feof(GIMPS)) {
  			ch = fgetc(GIMPS);
  			if(ch == '\n') {
    			gimpN++;
  			}
  		}

  		//fclose(GIMPS);
	  	GIMPS = fopen(gimpfile, "r");
	  	int i = 1;
	  	for (i = 1; i <= gimpN; i++) {
			fgets(gimp[i], sizeof(gimp[i]), GIMPS);
			memset(gimp[i]+strlen(gimp[i])-1, '\0', 1);
		}
	} else {
		printf("gimps.txt not found.\n");
		getchar();
		exit(1);
		gimpN = -1;
	}

	if (gimpN < 1) {
		printf("gimps.txt is empty!");
		getchar();
		exit(1);
	}

	//RegisterHotKey(NULL, 100, MOD_ALT | MOD_CONTROL, 'S');

	UINT hk = NULL;
    
	RegisterHotKey(NULL, 1, hk, 0x2D);
	RegisterHotKey(NULL, 2, hk, modekey);
	RegisterHotKey(NULL, 3, hk, 0x23);
	RegisterHotKey(NULL, 4, hk, 0xA5);
	

	MSG msg;
	srand(time(NULL));
	HKL kl = LoadKeyboardLayout("US", KLF_ACTIVATE);

	int delLength = 0;
	int gimpLog = 1;
	int mode = 1;
	/*char htest;
        printf("enter: ");
	scanf("%c", &htest);
	//short vret = VkKeyScanEx('I', kl);
	printf("\ngot '%hd'",vret);*/
	/*
	HANDLE h;
   
  if (!OpenClipboard(NULL))
    BailOut("Can't open clipboard");
   
  h = GetClipboardData(CF_TEXT);
 
  printf("%s\n", (char *)h);
 
  CloseClipboard();*/

	while (GetMessageA(&msg, NULL, 0, 0) != 0) {//main loop
		if (msg.wParam == 3) {//delete gimp (end key)
			int i = 0;
			for (i = 0; i < delLength + 1; i++) {
				keybd_event(0x08,0,0,0);
				keybd_event(0x08,0,KEYEVENTF_KEYUP,0);
			}
		}
		if (msg.wParam == 2) {//mode switch (home key)
			mode++;
			if (mode == 5) {
				mode = 1;
			}
			switch (mode) {
				case 1:
					printf("Switched to: normal random mode (1)\n");
					break;
				case 2:
					printf("Switched to: slow type mode (2)\n");
					break;
				case 3:
					printf("Switched to: top to bottom mode (3)\n");
					break;
				case 4:
					printf("Switched to: Gimp collection mode (4)\n");
					break;
			}
		}
		if (msg.wParam == 1) {//action (insert key)
			if (mode == 4) {//gimp collection mode
				
				HANDLE h;
   
				if (!OpenClipboard(NULL)) {
					BailOut("Can't open clipboard");
				}
				h = GetClipboardData(CF_TEXT);
				printf("\nPutting gimp \"%s\" into gimps.txt.\n\n", (char *)h);
				CloseClipboard();
				
				/*GIMPS = fopen("gimps.txt", "w");
				fprintf(GIMPS, "%s",(char*)h);
				fclose(GIMPS);
				FILE *GIMPS2;
				int gimpN2;
				*/
				FILE *GIMPS2;
				//int gimpN2;
				GIMPS2 = fopen("gimps.txt", "a+");
				//char gimp2[999999][512];
				
				if (!fileExists("gimps.txt")) {	
					printf("gimps.txt not found!\n");
					exit(1);
				}
				
				fprintf(GIMPS2, "\n%s", (char*)h);
				fclose(GIMPS2);
				
			}
			if (mode == 3) {//start to bottom mode
				 int i = 0;
				 char buf[512];
				 strcpy(buf, gimp[gimpLog]);
				 buf[strlen(buf)] = ' ';
				 delLength = strlen(gimp[gimpLog]);
				 for (i = 0; i < strlen(gimp[gimpLog]) + 1; i++) {
					if (buf[i] == '?' || buf[i] == '>' || buf[i] == ':' || buf[i] == '!' || buf[i] == '@' || buf[i] == '#' || buf[i] == '$' || buf[i] == '%' || buf[i] == '^' || buf[i] == '&' || buf[i] == '*' || buf[i] == '(' || buf[i] == ')' || buf[i] == '_' || buf[i] == '+' || buf[i] == '"' || buf[i] == '<' || buf[i] == '|' || buf[i] == '~' || isupper(buf[i]) != 0) {
						keybd_event(0x10,0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
						keybd_event(0x10,0,KEYEVENTF_KEYUP,0);
					} else {
						keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
					}
				 }
				 
				if (gimpLog == gimpN) { gimpLog = 0; }
				gimpLog++;
			}
			if (mode == 2) { //slow type
				 int randomnumber = rand() % (gimpN - 1) + 1;
				 int i = 0;
				 char buf[512];
				 strcpy(buf, gimp[randomnumber]);
				 buf[strlen(buf)] = ' ';
				 delLength = strlen(gimp[randomnumber]);
				 for (i = 0; i < strlen(gimp[randomnumber]) + 1; i++) {
					if (buf[i] == '?' || buf[i] == '>' || buf[i] == ':' || buf[i] == '!' || buf[i] == '@' || buf[i] == '#' || buf[i] == '$' || buf[i] == '%' || buf[i] == '^' || buf[i] == '&' || buf[i] == '*' || buf[i] == '(' || buf[i] == ')' || buf[i] == '_' || buf[i] == '+' || buf[i] == '"' || buf[i] == '<' || buf[i] == '|' || buf[i] == '~' || isupper(buf[i]) != 0) {
						keybd_event(0x10,0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
						keybd_event(0x10,0,KEYEVENTF_KEYUP,0);
					} else {
						keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
						keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
					}
					Sleep(80); //60 before, 90 after, 75 now?
				 }
				 keybd_event(0x0D,0,0,0);
			}
			if (mode == 1) {//regular
				int randomnumber = rand() % (gimpN - 1) + 1;
             			int i = 0;
             			char buf[512];
				strncpy(buf, gimp[randomnumber], strlen(gimp[randomnumber]));
             			delLength = strlen(gimp[randomnumber]);
             			for (i = 0; i < strlen(gimp[randomnumber]); i++) {
             				if (buf[i] == '?' || buf[i] == '>' || buf[i] == ':' || buf[i] == '!' || buf[i] == '@' || buf[i] == '#' || buf[i] == '$' || buf[i] == '%' || buf[i] == '^' || buf[i] == '&' || buf[i] == '*' || buf[i] == '(' || buf[i] == ')' || buf[i] == '_' || buf[i] == '+' || buf[i] == '"' || buf[i] == '<' || buf[i] == '|' || buf[i] == '~' || isupper(buf[i]) != 0) {
             					keybd_event(0x10,0,0,0);
             					keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
             					keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
             					keybd_event(0x10,0,KEYEVENTF_KEYUP,0);
             				} else {
             					keybd_event(VkKeyScanEx(buf[i], kl),0,0,0);
             					keybd_event(VkKeyScanEx(buf[i], kl),0,KEYEVENTF_KEYUP,0);
             				}
              			}
	         		keybd_event(VkKeyScanEx(' ', kl),0,0,0); //print space at the end of gimp
             			keybd_event(VkKeyScanEx(' ', kl),0,KEYEVENTF_KEYUP,0);

			}
		}

		//delLength = strlen(gimp[randomnumber]);
	}

	return 0;
} 
